[DOCU-2617] Secrets management in konnect (#4828)
Kong/docs.konghq.com: app/konnect/runtime-manager/vaults/index.md

---
title: Secrets Management in Konnect
content_type: explanation
---

> Secure your {{site.konnect_short_name}} environment with centralized secrets,
> making it easier to manage and build governance.

Reviewer: ```suggestion
Secrets management in {{site.konnect_short_name}} allows you to store secrets in centralized vaults, making it easier to manage security and governance policies. 
```

> Secrets may be part of the core gateway configuration,

Reviewer: ```suggestion
Secrets can be part of the core gateway configuration,
```

or part of gateway configuration associated with APIs serviced by the gateway.

> Some of the most common types of secrets include:

Reviewer: ```suggestion
The most common types of secrets include:
```

* Certificates
* API keys
* Personal access tokens
* Credentials for databases
* Certain plugin fields, like `session_secret` in the OIDC plugin

You can use vaults to safely store and retrieve secrets used in {{site.base_gateway}}
deployments, improving the fundamental security of your applications.

> In configuration, you can then reference the secrets stored in vaults instead

Reviewer: ```suggestion
In the configuration, you can reference the secrets stored in vaults as variables instead
```

> of displaying the actual value of the secret.

Reviewer: ```suggestion
of displaying the actual value of the secret in plaintext. This way, the {{site.konnect_short_name}}
platform never stores sensitive credentials. 
```

> This ensures that secrets are not stored in plaintext throughout the platform,
> in places such as `kong.conf` files, in declarative configuration files, logs,
> or in the {{site.konnect_short_name}} UI. This way, the {{site.konnect_short_name}}
> platform never stores sensitive credentials. 

Reviewer: ```suggestion

```

---
title: Secrets Management in Konnect
content_type: explanation
---

Secure your {{site.konnect_short_name}} environment with centralized secrets,
making it easier to manage and build governance.

> A secret is a sensitive piece of information required for proper gateway operations.

Reviewer: ```suggestion

```

Secrets may be part of the core gateway configuration,
or part of gateway configuration associated with APIs serviced by the gateway.
Some of the most common types of secrets include:
* Certificates
* API keys
* Personal access tokens
* Credentials for databases
* Certain plugin fields, like `session_secret` in the OIDC plugin

You can use vaults to safely store and retrieve secrets used in {{site.base_gateway}}
deployments, improving the fundamental security of your applications.
In configuration, you can then reference the secrets stored in vaults instead
of displaying the actual value of the secret.

This ensures that secrets are not stored in plaintext throughout the platform,
in places such as `kong.conf` files, in declarative configuration files, logs,
or in the {{site.konnect_short_name}} UI. This way, the {{site.konnect_short_name}}

> platform never stores or sees sensitive credentials, letting you use the security controls

Reviewer: ```suggestion
platform never stores sensitive credentials. 
```

> that you already have around sensitive information.

Reviewer: ```suggestion
```

## Vaults interface in {{site.konnect_short_name}}

![Vaults interface](/assets/images/docs/konnect/konnect-vaults.png)
> Figure 1: Overview page for all vaults configured for a runtime group.

Number | Item | Description
-------|------|------------
1 | **Vaults menu link** | Main link to the vaults configuration for a runtime group. Appears when you select a runtime group.
2 | **Add vault** | Click the **Add vault** button to set up any supported Konnect vault backend.

> 3 | **Table entries for vaults** | Select a table entry to see the configuration for a particular vault. On each vault's configuration page, you can edit or delete the vault, or copy the entire configuration as JSON.

Reviewer: ```suggestion
3 | **Vault entry** | Select a vault entry to open the configuration page for the particular vault. On each vault's configuration page, you can edit or delete the vault, or copy the entire configuration as JSON.
```

> 4 | **Vault ID** | The vault's UUID. You can click the copy icon to copy the ID for use in configuration.

Reviewer: ```suggestion
4 | **Vault ID** | The vault's UUID. 
```

Author: Hmm. I'd like to leave that out for now. I'm not 100% sure why you'd need to do that at the moment, because there's no way to edit it through the API (I think?). That would be the main reason to use a vault UUID.

> 5 | **Vault action menu** | Interact with a configured vault. From this menu, you can <b>view</b>, <b>edit</b>, or <b>delete</b> a vault's configuration. <br><br> <b>Be careful when deleting a configured vault!</b> Once deleted, a vault can't be restored, and all vault references will fail if not updated.

Reviewer: ```suggestion
5 | **Vault action menu** | From this menu, you can <b>view</b>, <b>edit</b>, or <b>delete</b> a vault's configuration. 
```

Reviewer: I think the deleted part also applies to the information from within the configuration menu. If there is no UI text that appears that states that info when you try to delete a vault, then we should just add that information under the table in a note.

Author: There is UI text for this, so i'll just remove.

> ## Benefits

Reviewer: ```suggestion
## Use cases
```

> Use vaults for the following benefits:

Reviewer: ```suggestion
Vaults have several use cases: 
```

* Storing secrets securely

> * Managing access to secrets with fine grained policies

Reviewer: ```suggestion
* Managing access to secrets with fine-grained policies
```

> * Automation of secrets rotation

Reviewer: ```suggestion
* Automating the rotation of secrets.
```

You can ignore this one.

> * Audits of secrets usage

Reviewer: ```suggestion
* Auditing secrets usage
```

* Encryption of secrets at rest

> Be aware that {{site.konnect_short_name}} **does not**:

Reviewer: ```suggestion
{{site.konnect_short_name}} **does not**:
```

> * Store credentials to access the vault itself.

Reviewer: ```suggestion
* Store credentials to access the vault itself.
```

How about *pass credentials to the data plane

Author: I would argue that the "storing" part is more important to get across. It doesn't have any credentials to pass. Konnect doesn't show or store your credentials anywhere; only the data plane gets access to the credentials.

You must provide those credentials to the {{site.base_gateway}} data plane directly.

> * Update or modify the secrets in the 3rd party vaults.

Reviewer: ```suggestion
* Update or modify the secrets in 3rd party vaults.
```

Vaults are configurable per runtime group. You can't use the same vault across
multiple runtime groups.

## Supported vaults in {{site.konnect_short_name}}

> Konnect supports the following vault backends:

Reviewer: Should the order for these have environment variables at the bottom?

> * Environment variables
> * AWS Secrets Manager
> * HashiCorp Vault
> * GCP Secret Manager

Author: ```suggestion
* AWS Secrets Manager
* HashiCorp Vault
* GCP Secret Manager
* Environment variables
```

> You can manage all of these vaults through the [Konnect UI](/konnect/runtime-manager/vaults/how-to) or with [decK](/deck/latest/guides/vaults/).

Reviewer: I think this should say "..through the Konnect Manager", or "..through the Konnect web application". We use Konnect UI a lot, but I think using UI here implies that it's just a UI wrapping around something, and I think we want to position Konnect own standalone thing.

Author: I assume there will eventually be an API? But I can say through the Runtime Manager in Konnect.

> You can manage all of these vaults through the [Konnect UI](/konnect/runtime-manager/vaults/how-to) or with [decK](/deck/latest/guides/vaults/).

Author: ```suggestion
You can manage all of these vaults through the [Runtime Manager](/konnect/runtime-manager/vaults/how-to) or with [decK](/deck/latest/guides/vaults/).
```
